Multi-option support in custom benchmarks

deanstef opened the pull request:
This PR adds CLI support to specify multiple options over `custom` benchmark runs.

deanstef committed:
multi-option support in custom benchmarks

dkgoutham committed:
Merge branch 'main' into bench/multiple-options

dkgoutham committed:
Fixed code quality related issues, and completed multiple parameter feature and also added multiple field option

dkgoutham commented:
@Oghma Can you please review? I have addressed the code quality-related issues, and have also included the multiple parameters option for field (f64, f128) as well in the CLI bench.

Oghma commented on bench/src/defrida.rs:
```diff
@@ -419,74 +419,108 @@ pub fn run_full_benchmark(output_path: &str) {
-    pub num_validators: usize,
-    pub num_queries: usize,
-    pub batch_size: usize,
+    pub fri_options: Vec<(usize, usize, usize)>,
```
Why use usize? Use `usize` for numbers isn't correct. 
`usize` type are meant to deal with memory access related operations (array indexing, poiner arithmetic, type punning with pointers, ffi interop, etc). if you use it like a number, you are probably mis-using it.

Oghma commented on bench/src/defrida.rs:
```diff
@@ -419,74 +419,108 @@ pub fn run_full_benchmark(output_path: &str) {
-            field_names::F128,
-        );
-        results.push(result_f128);
+    for &(blowup_factor, folding_factor, max_remainder_degree) in &config.fri_options {
```
We have to reduce the amount of nested loops. There are functions like`itertools::cartesian_product`, `flat_map` etc. Also [`iproduct!`](https://docs.rs/itertools/latest/itertools/macro.iproduct.html)

Oghma commented on bench/src/defrida.rs:
```diff
@@ -419,74 +419,108 @@ pub fn run_full_benchmark(output_path: &str) {
+            for &num_validators in &config.num_validators {
+                for &num_queries in &config.num_queries {
+                    for &batch_size in &config.batch_sizes {
+                        let options =
```
Move this line between the first and the first nested loop. Line 436. No need to create it every time

Oghma commented on bench/src/frida.rs:
```diff
@@ -404,68 +404,103 @@ pub fn run_full_benchmark(output_path: &str) {
-    let options = FriOptions::new(blowup_factor, folding_factor, max_remainder_degree);
-    let mut results = Vec::new();
+pub struct CustomFridaBenchmarkConfig<'a> {
+    pub fri_options: Vec<(usize, usize, usize)>,
```
Same as `defrida.rs`

Oghma commented on bench/src/defrida.rs:
```diff
@@ -419,74 +419,108 @@ pub fn run_full_benchmark(output_path: &str) {
+    pub num_validators: Vec<usize>,
+    pub num_queries: Vec<usize>,
+    pub batch_sizes: Vec<usize>,
+    pub field_type: crate::common::FieldType,
```
The `FieldType` is used a lot. Consider importing it.

Oghma commented on bench/src/frida.rs:
```diff
@@ -404,68 +404,103 @@ pub fn run_full_benchmark(output_path: &str) {
+    pub data_sizes: Vec<usize>,
+    pub batch_sizes: Vec<usize>,
+    pub num_queries: Vec<usize>,
+    pub field_type: crate::common::FieldType,
```
The `FieldType` is used a lot. Consider importing it.

Oghma commented on bench/src/frida.rs:
```diff
@@ -404,68 +404,103 @@ pub fn run_full_benchmark(output_path: &str) {
-            field_names::F128,
-        );
-        results.push(result_f128);
+    for &(blowup_factor, folding_factor, max_remainder_degree) in &config.fri_options {
```
We need to think about how to abstract and remove this code. It's a copy and paste from `defrida.rs`.

dkgoutham replied:
I kind of agree, but its not fully the same code. But yea, I will think of something!

dkgoutham replied:
I have been thinking of a trait based approach, have been working, will push it

Oghma commented on bench/src/main.rs:
```diff
@@ -41,18 +42,16 @@ enum BenchmarkSubcommand {
-        num_queries: usize,
+        #[arg(long, default_value = "(2,2,0)")]
+        fri_options: String,
+        #[arg(long, use_value_delimiter = true, default_value = "65536")]
```
Is the default value correct?

dkgoutham replied:
Default value could be anything honestly, (2,2,0) is the most commonly used so that's the reason I went with that! It's basically to provide a better UX when using the cli

Oghma replied:
I'm pointing at `default_value = "65536"`. Is it correct that the vector will be created with 65.5k integers by default?

dkgoutham replied:
No, a vector of single element, with value 65536 (2^16) will be created

Oghma commented on bench/src/main.rs:
```diff
@@ -65,16 +64,14 @@ enum SingleFridaSubcommand {
-        batch_size: usize,
+        #[arg(long, default_value = "(2,2,0)")]
+        fri_options: String,
+        #[arg(long, use_value_delimiter = true, default_value = "65536")]
```
Same as above

dkgoutham committed:
Refactored nested loops with iproduct macro

dkgoutham commented:
Used [iproduct! macro](https://docs.rs/itertools/latest/itertools/macro.iproduct.html) instead of nested loops!

dkgoutham reviewed (commented):
Added the field option feature as well!

Oghma commented on bench/src/main.rs:
```diff
@@ -41,18 +42,16 @@ enum BenchmarkSubcommand {
-        batch_size: usize,
-        #[arg(long, default_value = "32")]
-        num_queries: usize,
+        #[arg(long, default_value = "(2,2,0)")]
```
What is the input? Is it a list of tuples? Are you sure that clap cannot handle it?

dkgoutham replied:
Yea list of tuples, I was getting some clap error initially, and hence went with the custom parse_fri_options() function.

Oghma commented on bench/README.md:
```diff
@@ -28,6 +28,7 @@ bench/
 ### Prerequisites
+
 - Rust toolchain with `cargo`
 - Feature flag `bench` enabled for compilation
```
Is it always required? If so, why not add to `default`?

dkgoutham replied:
Sure, can do it. It's just that I left things untouched, the part of the code that was written by the previous author.

Oghma replied:
Ok, leave it as it is.

Oghma commented on src/prover/builder.rs:
```diff
@@ -115,6 +115,11 @@ where
     fn prepare_prover_state(&self, data: &[u8], num_queries: usize) -> ProverStateResult<E, H> {
+        #[cfg(feature = "bench")]
+        unsafe {
+            bench::TIMER = Some(Instant::now());
```
Why do you need a static variable?

dkgoutham replied:
You mean the bench::TIMER? Previous author code, didn't go to change it, its related to frida benchmarking, which was initially written 1 year ago, but I modified it to our template of CLI, but didn't change this bench::TIMER, maybe we can have this as an issue, which we can address in the future?

dkgoutham committed:
Created new file runner.rs and wrote Benchmark trait, and then refacto
red code from frida.rs, defrida.rs and single_frida.rs and updated the main.rs accordingly

dkgoutham committed:
formated the code

dkgoutham commented:
- runner.rs file was introduced, where new trait and 2 structs for parameters have been introduced, and this is were run_benchmark code with iproduct! resides.
- frida.rs, defrida.rs and single_frida.rs now implement this trait, and thereby abstracting away code duplication.
- Modified the main.rs accordingly.

Need to address usize, will do after @Oghma reviews this.

dkgoutham committed:
Updated the bench/README.md file

dkgoutham committed:
Made changes from usize to u32/u64 wherever appropriate

dkgoutham commented:
The thing is, `winter-fri` crate (a dependency) uses `usize` in it's APIs. So I used appropriate `u32` and `u64` types within our benchmark logic and then cast them to `usize` only when calling `winter-fri` function that requires it. Same way, when we get a `usize` back from a winter-fri function, casted it to the appropriate specific type.

Oghma reviewed (approved):
LGTM

dkgoutham commented:
Can we merrge this? @maksimryndin

maksimryndin reviewed (approved):
LGTM! 👍

dkgoutham commented:
Can we relax the "commits must have verified signatures" check please? Initial two commits by Stefano are not verified, so let's relax that for this PR only! @maksimryndin
